src/main/java/wooteco/subway/domain/Sections.java

```diff
@@ -0,0 +1,225 @@
+    private static final int OVERLAP_COUNT_ADD_MIDDLE = 2;
+
+    private final List<Section> values;
+
+    public Sections(List<Section> sections) {
+        values = new LinkedList<>(sortedSections(sections));
+    }
+
+    private List<Section> sortedSections(List<Section> sections) {
+        if (sections.size() <= 1) {
+            return sections;
+        }
+        Long upDestinationId = findUpDestinationId(sections);
+        Map<Long, Section> sectionsByIds = sections.stream()
+            .collect(Collectors.toMap(Section::getUpStationId, section -> section, (a, b) -> b));
+
+        return getSortedValuesBy(upDestinationId, (key) -> sectionsByIds.get(key).getDownStationId(), sectionsByIds);
+    }
+
+    private Long findUpDestinationId(List<Section> sections) {
+        List<Long> upStations = collectValuesBy(Section::getUpStationId, sections);
+        List<Long> downStations = collectValuesBy(Section::getDownStationId, sections);
+
+        upStations.removeAll(downStations);
+        if (upStations.size() != 1) {
+            throw new IllegalStateException("상행 종점을 찾을 수 없습니다. 구간 목록이 올바르게 저장되지 않습니다.");
+        }
+        return upStations.get(0);
+    }
+
+    private <T, E> List<T> collectValuesBy(Function<E, T> mapper, List<E> collection) {
+        return collection.stream()
+            .map(mapper)
+            .collect(Collectors.toList());
+    }
+
+    private <T> List<T> getSortedValuesBy(Long firstKey, Function<Long, Long> newKeyMapper, Map<Long, T> values) {
+        List<T> result = new ArrayList<>();
+        Long key = firstKey;
+
+        while (values.containsKey(key)) {
+            result.add(values.get(key));
+            key = newKeyMapper.apply(key);
+        }
+
+        return result;
+    }
+
+    public Section add(Section newSection) {
+        validateAddable(newSection);
+
+        if (getUpDestination().equals(newSection.getDownStation())) {
+            values.add(0, newSection);
+            return newSection;
+        }
+
+        if (getDownDestination().equals(newSection.getUpStation())) {
+            values.add(newSection);
+            return newSection;
+        }
+
+        for (Section section : values) {
+            if (section.hasSameUpStation(newSection)) {
+                return insertAfterUpStation(newSection, section);
+            }
+
+            if (section.hasSameDownStation(newSection)) {
+                return insertBeforeDownStation(newSection, section);
+            }
+        }
+        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
+    }
+
+    private Section insertAfterUpStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), newSection.getDownStation(), section.getDownStation(),
+            getNewDistance(section, newSection));
+        values.set(index, newSection);
+        values.add(index + 1, sectionToModify);
+        return sectionToModify;
+    }
```

Comment:
insertStation 메서드를 만들고 메서드 안에서 분기처리하면 add메서드에서 분기처리하지 않을 수 있겠어요
insertAfterUpStation메서드와 insertBeforeDownStation메서드는 구조가 비슷해보여서요!

Reply:
```suggestion
        for (Section section : values) {
            Optional<Section> insertedSection = insertStation(newSection, section);
            if (insertedSection.isPresent()) {
                return insertedSection.get();
            }
        }

        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
    }

    private Optional<Section> insertStation(Section newSection, Section section) {
        if (section.hasSameUpStation(newSection)) {
            return Optional.of(insertAfterUpStation(newSection, section));
        }
        if (section.hasSameDownStation(newSection)) {
            return Optional.of(insertBeforeDownStation(newSection, section));
        }
        return Optional.empty();
    }
```

얼리 리턴때문에 메서드를 분리함면 null 을 반환해줘야 하는 상황이 생겨서 Optional 을 사용해서 분리해보았습니다 !!
로직이 깔끔해지긴 한 것 같은데, `insertStation` 메서드만 보았을 때 추가된 Station 을 받는게 명확하게 파악되지 않을 것 같다는 걱정도 되네요. 😹 일단 최선의 방법으로 선택해보았습니다!! 제안해주셔서 감사합니다. 😆👍🏻👍🏻👍🏻👍🏻

Reply:
아.. `throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");` 이 코드가 for문 밖에 있었군요..
for문 안에 있다고 생각해서 insertStation 메서드 안에서 예외를 발생시키도록하면 add메서드를 정리할 수 있을줄 알았어요..
라라 생각대로 insertStation 메서드가 살짝 모호해졌네요.  변경전 코드가 더 나을 수 있겠어요. 😹

그래도 최대한 반영해주셔서 감사합니다!